SAL annotations for wil::mapview_deleter::operator() are incorrect

The current implementation generates an error when running code analysis on code that uses wil::unique_mapview_ptr<T>.

In particular the annotations on mapview_deleter::operator() are incorrect it currently has (_Pre_opt_valid _Frees_ptr_opt) however that doesn't match the requirements and declaration of UnMapViewOfFile, which doesn't accept NULL (e.g. the "opt" portion of the annotations is erroneous))

FYI the error is:
`packages\Microsoft.Windows.ImplementationLibrary.1.0.200902.2\include\wil\resource.h(3456): warning C6387: 'p' could be '0':  this does not adhere to the specification for the function 'UnmapViewOfFile'. `

--- a/include/wil/resource.h
+++ b/include/wil/resource.h
@@ -1298,10 +1298,7 @@ namespace wil
     {
         template<typename T> void operator()(_Frees_ptr_opt_ T* toFree) const
         {
-            if (toFree)
-            {
-                TDeleter(toFree);
-            }
+            TDeleter(toFree);
         }
     };
 

--- a/include/wil/resource.h
+++ b/include/wil/resource.h
@@ -3438,7 +3438,7 @@ namespace wil
     struct process_heap_deleter
     {
         template <typename T>
-        void operator()(_Pre_opt_valid_ _Frees_ptr_opt_ T* p) const
+        void operator()(_Pre_valid_ _Frees_ptr_ T* p) const
         {
             details::FreeProcessHeap(p);
         }
@@ -3447,7 +3447,7 @@ namespace wil
     struct virtualalloc_deleter
     {
         template<typename T>
-        void operator()(_Pre_opt_valid_ _Frees_ptr_opt_ T* p) const
+        void operator()(_Pre_valid_ _Frees_ptr_ T* p) const
         {
             ::VirtualFree(p, 0, MEM_RELEASE);
         }
@@ -3456,7 +3456,7 @@ namespace wil
     struct mapview_deleter
     {
         template<typename T>
-        void operator()(_Pre_opt_valid_ _Frees_ptr_opt_ T* p) const
+        void operator()(_Pre_valid_ _Frees_ptr_ T* p) const
         {
             ::UnmapViewOfFile(p);
         }
@@ -3489,7 +3489,7 @@ namespace wil
     /** Manages a typed pointer allocated with MapViewOfFile
     A specialization of wistd::unique_ptr<> that frees via UnmapViewOfFile(p).
     */
-    template<typename T>
+    template<typename T = void>
     using unique_mapview_ptr = wistd::unique_ptr<T, mapview_deleter>;
 
 #endif // __WIL_WINBASE_

--- a/include/wil/resource.h
+++ b/include/wil/resource.h
@@ -1298,7 +1298,10 @@ namespace wil
     {
         template<typename T> void operator()(_Frees_ptr_opt_ T* toFree) const
         {
-            TDeleter(toFree);
+            if (toFree)
+            {
+                TDeleter(toFree);
+            }
         }
     };
 

--- a/include/wil/result.h
+++ b/include/wil/result.h
@@ -406,10 +406,9 @@ namespace wil
 
                 if (shouldAllocate)
                 {
-                    Node *pNew = reinterpret_cast<Node *>(details::ProcessHeapAlloc(0, sizeof(Node)));
-                    if (pNew != nullptr)
+                    if (auto pNewRaw = details::ProcessHeapAlloc(0, sizeof(Node)))
                     {
-                        new(pNew)Node{ threadId };
+                        auto pNew = new (pNewRaw) Node{ threadId };
 
                         Node *pFirst;
                         do

--- a/include/wil/win32_helpers.h
+++ b/include/wil/win32_helpers.h
@@ -15,7 +15,6 @@
 #include <sysinfoapi.h> // GetSystemTimeAsFileTime
 #include <libloaderapi.h> // GetProcAddress
 #include <Psapi.h> // GetModuleFileNameExW (macro), K32GetModuleFileNameExW
-#include <PathCch.h>
 #include <objbase.h>
 
 #include "result.h"

--- a/include/wil/winrt.h
+++ b/include/wil/winrt.h
@@ -19,6 +19,7 @@
 #include "result.h"
 #include "com.h"
 #include "resource.h"
+#include <windows.foundation.h>
 #include <windows.foundation.collections.h>
 
 #ifdef __cplusplus_winrt
@@ -59,7 +60,6 @@ namespace std
 
 namespace wil
 {
-#ifdef _INC_TIME
     // time_t is the number of 1 - second intervals since January 1, 1970.
     long long const SecondsToStartOf1970 = 0x2b6109100;
     long long const HundredNanoSecondsInSecond = 10000000LL;
@@ -76,7 +76,6 @@ namespace wil
         dateTime.UniversalTime = (timeT + SecondsToStartOf1970) * HundredNanoSecondsInSecond;
         return dateTime;
     }
-#endif // _INC_TIME
 
 #pragma region HSTRING Helpers
     /// @cond

--- a/tests/ResourceTests.cpp
+++ b/tests/ResourceTests.cpp
@@ -740,4 +740,5 @@ TEST_CASE("DefaultTemplateParamCompiles", "[resource]")
 
     wil::unique_midl_ptr<> g;
     wil::unique_cotaskmem_ptr<> h;
+    wil::unique_mapview_ptr<> i;
 }

--- a/tests/WinRTTests.cpp
+++ b/tests/WinRTTests.cpp
@@ -1,5 +1,4 @@
 
-#include <ctime> // TODO: https://github.com/microsoft/wil/issues/44
 #include <wil/winrt.h>
 
 #ifdef WIL_ENABLE_EXCEPTIONS

